src/v/kafka/server/snc_quota_manager.cc

```diff
@@ -297,10 +298,103 @@ void snc_quota_manager::maybe_arm_balancer_timer() {
     _balancer_timer.arm(arm_until);
 }
 
+using shard_count_t
+  = boost::uint_t<std::numeric_limits<ss::shard_id>::digits>::fast;
+
+struct borrower_t {
+    ss::shard_id shard_id;
+    // can be 0 or 1, made shard_count_t for easier aggregation
+    inoutpair<shard_count_t> is_borrower;
+};
+
 ss::future<> snc_quota_manager::quota_balancer_step() {
     const auto sem_units = co_await ss::get_units(_balancer_semaphore, 1);
     _balancer_gate.check();
     _balancer_timer_last_ran = ss::steady_clock_type::now();
+    vlog(klog.trace, "qb - Step");
+
+    // determine the borrowers and whether any balancing is needed now
+    const auto now = clock::now();
+    const std::vector<borrower_t> borrowers = co_await container().map(
```

Comment:
`const auto .. = `?

Reply:
If everyone is aware that `map()` retuns a `vector` and not any other container, that may be ok. I'm trying to resort to `auto` only when it's crystal clear from the nearby code what the type is ... and in templates :)

Reply:
i think it's safe to say that map returning a vector is common information.